Refactoring of cmake structure

BSolonenko opened the pull request:
SDL may use modern cmake approach of creating targets. It will simplify
porting SDL on any platform. For each third party library used by SDL
created .cmake file, that will expose library as cmake_library.

This will allow to make clearer dependencies, avoid dependency gaps
(required for multithreading compilation) and replace dependency during
porting to other platform just with replacing .cmake third party library
file.

LuxoftAKutsan commented on CMakeLists.txt:
```diff
@@ -224,7 +280,6 @@ add_custom_target(genivi-tarball
 
-project (${PROJECT})
 
 #ADD_DEPENDENCIES(${PROJECT} Policy)
```
@BSolonenko remove commented code

LuxoftAKutsan commented on CMakeLists.txt:
```diff
@@ -333,24 +388,6 @@ if (TELEMETRY_MONITOR)
   message(STATUS "Telemetry monitor enabled")
 endif()
 
-if (${EXTENDED_POLICY} STREQUAL "EXTERNAL_PROPRIETARY")
```
@BSolonenko did policy management moved somewhere?

LuxoftAKutsan commented on src/appMain/CMakeLists.txt:
```diff
@@ -28,231 +28,226 @@
+
+target_sources(${PROJECT_NAME}
+  PRIVATE
+    $<$<PLATFORM_ID:QNX>:${CMAKE_CURRENT_LIST_DIR}/qnx_main.cc>
```
@BSolonenko  we have no qnx_main.cc in open source

LuxoftAKutsan commented on src/appMain/CMakeLists.txt:
```diff
@@ -28,231 +28,226 @@
+    transport_manager::transport_manager
+    $<$<BOOL:${TELEMETRY_MONITOR}>:telemetry_monitor::telemetry_monitor>
+    utils::utils
+    $<$<BOOL:${SYNC_USE_MQ_SERVER}>:mq>
```
@BSolonenko we have no MQ in open source

LuxoftAKutsan commented on src/appMain/CMakeLists.txt:
```diff
@@ -28,231 +28,226 @@
+  PRIVATE
+    components::interface
+    Policy::usage_statistics
+    libboost_filesystem.so.1.66.0
```
@BSolonenko do we have ability to use modern cmake approach for boost libraires? 
like boost::filesystem ?

LuxoftAKutsan commented on src/appMain/CMakeLists.txt:
```diff
@@ -28,231 +28,226 @@
-    list(APPEND LIBRARIES bluetooth)
-  endif()
-endif()
+# if (BUILD_BT_SUPPORT)
```
@BSolonenko why commented?

LuxoftAKutsan commented on src/appMain/CMakeLists.txt:
```diff
@@ -28,231 +28,226 @@
 if (BUILD_USB_SUPPORT)
-  if (CMAKE_SYSTEM_NAME STREQUAL "Linux")
-    list(APPEND LIBRARIES libusb-1.0.so)
+  if(UNIX AND NOT QNXNTO)
```
@BSolonenko can we use cmake generators for it?

LuxoftAKutsan commented on src/appMain/CMakeLists.txt:
```diff
@@ -28,231 +28,226 @@
+
+  # Install log properties for debug reasons.
+  # TODO: Remove this for production
+  set(SDL_IN_LOG_CONFIG_FILE "${CMAKE_CURRENT_SOURCE_DIR}/log4cxx_qnx.debug.properties")
```
@BSolonenko we should not have and qnx logger configs.

LuxoftAKutsan commented on src/components/CMakeLists.txt:
```diff
@@ -28,66 +28,65 @@
-  add_subdirectory(./policy/policy_external/)
-  message(STATUS "Use external policy")
+set(COMPONENTS_DIR ${CMAKE_CURRENT_LIST_DIR})
+project(components)
```
@BSolonenko do we rely have project components? what is the sources of this project? 
If it just collects all the components, why don't we use this project for linkage of appMain?

LuxoftAKutsan commented on src/components/application_manager/CMakeLists.txt:
```diff
@@ -31,156 +31,125 @@
+    hmi_message_handler::hmi_message_handler
+    Policy::Policy
+    Policy::usage_statistics
+    bson
```
@BSolonenko can we connect bson in modern cmake approach ?

LuxoftAKutsan commented on src/components/application_manager/rpc_plugins/sdl_rpc_plugin/CMakeLists.txt:
```diff
@@ -28,50 +28,364 @@
+    # TODO: Temporary solution for fixing error:
+    # 'LoadPlugin: Failed to open dll .//libsdl_rpc_plugins.so : .//libsdl_rpc_plugins.so: undefined symbol: _ZN5utils11version_4_5E'
+    # Provide symbol 'version_4_5' for plugin
+    "${CMAKE_SOURCE_DIR}/src/components/utils/src/semantic_version.cc"
```
@BSolonenko what solution would be not temporary ?

LuxoftAKutsan commented on src/components/connection_handler/CMakeLists.txt:
```diff
@@ -28,31 +28,30 @@
+    protocol::protocol
+    utils::utils
+    protocol_handler::protocol_handler
+    #transport_manager::transport_manager
```
@BSolonenko why commented?

LuxoftAKutsan commented on src/components/hmi_message_handler/CMakeLists.txt:
```diff
@@ -28,59 +28,49 @@
+  PUBLIC
+    utils::utils
+    application_manager::application_manager
+    $<$<BOOL:${SYNC_USE_MQ_SERVER}>:mq>
```
@BSolonenko please remove all mq

LuxoftAKutsan commented on src/components/policy/policy_external/CMakeLists.txt:
```diff
@@ -29,94 +29,60 @@
-  ${POLICY_PATH}/src/sql_pt_ext_representation.cc
+target_link_libraries(${PROJECT_NAME}
+  PUBLIC
+    jsoncpp
```
@BSolonenko please use modern cmake approach for it

LuxoftAKutsan commented on src/components/policy/policy_regular/CMakeLists.txt:
```diff
@@ -1,120 +1,55 @@
-# Copyright (c) 2016, Ford Motor Company
```
@BSolonenko copyright should present

LuxoftAKutsan commented on tools/InterfaceGenerator/test/generator/generators/test_expected_sdlrpc.cc:
```diff
@@ -136,15 +136,15 @@ CSmartSchema XXX::YYY::ZZZ::Test::InitFunction_name1_request(
-  params_members[ns_smart_device_link::ns_json_handler::strings::S_CORRELATION_ID] = CObjectSchemaItem::SMember(TNumberSchemaItem<int>::create(), true);
+  params_members[NsSmartDeviceLink::NsJSONHandler::strings::S_FUNCTION_ID] = CObjectSchemaItem::SMember(TEnumSchemaItem<FunctionID::eType>::create(function_id_items), true);
+  params_members[NsSmartDeviceLink::NsJSONHandler::strings::S_MESSAGE_TYPE] = CObjectSchemaItem::SMember(TEnumSchemaItem<messageType::eType>::create(message_type_items), true);
+  params_members[NsSmartDeviceLink::NsJSONHandler::strings::S_PROTOCOL_VERSION] = CObjectSchemaItem::SMember(TNumberSchemaItem<int>::create(), true);
```
@BSolonenko please extract namespace replacment to separate commit

Commit:
Removed the old CMake tree

Commit:
CMake tree for 3rd_party was added

Commit:
CMake tree for 3rd_party-static was added

Commit:
CMake tree for application_manager was added

Commit:
CMake tree was added

CMake tree for config_profile, connection_handler, formatters,
hmi_message_handler, media_manager, protocol, protocol_handler,
resumption and for rpc_base was added.

Commit:
CMake tree for utils was added

Commit:
CMake tree for policy was added

Commit:
CMake tree was added

CMake tree for interfaces, security_manager, smart_objects,
telemetry_monitor, transport_manager and for plugins was added.

Commit:
Refactoring of cmake structure

Commit:
answer review comments
